Support extensions from OCI registries with self-signed CA

byDimasik opened the pull request:
## Description

<!-- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change. -->

Allow providing TLS certificates for OCI registries in Helm commands. Reuse existing v1beta1.Repository type to avoid changing CRDs.

While k0s already supports custom TLS transport for classic Helm repositories, it is lacking the same functionality for OCI registries. This makes it impossible to install Helm extensions from OCI registries that use private/self-signed CA for TLS.

With this change, users can add their OCI registry as a Helm repository and provide TLS certificates. This WILL NOT modify the repositories.yaml file, but instead create a custom registry client for the provided OCI registry. This client will be used with charts that reference the same registry in the chartName.

Fixes https://github.com/k0sproject/k0s/issues/5877

## Type of change

<!-- check the related options -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation update

## How Has This Been Tested?

- [ ] Manual test
- [x] Auto test added

<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration -->

## Checklist

- [x] My code follows the style [guidelines](https://github.com/k0sproject/k0s/blob/main/docs/contributors/overview.md) of this project
- [x] My commit messages are [signed-off](https://github.com/k0sproject/k0s/blob/main/docs/contributors/github_workflow.md)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings

twz123 commented on pkg/helm/oci.go:
```diff
@@ -0,0 +1,111 @@
+		return m.getDefaultRegistryClient()
+	}
+
+	client, exists := m.knownClients[registryURL.Host]
```
Did you consider to create all clients on the fly instead of caching them? We don't cache the default client, so why should we do it for the others?

byDimasik replied:
done

byDimasik commented on pkg/helm/oci.go:
```diff
@@ -0,0 +1,111 @@
+
+	registryClient, err := registry.NewClient(
+		registry.ClientOptWriter(os.Stdout),
+		registry.ClientOptEnableCache(true),
```
The only reason I enabled this is because it's enabled by default in helm - https://github.com/helm/helm/blob/main/pkg/cmd/root.go#L362

I'll add a comment

byDimasik replied:
done

byDimasik commented on pkg/helm/oci_test.go:
```diff
@@ -0,0 +1,210 @@
+	testKeyPath  = "./testdata/tls/tls.key"
+)
+
+func TestOciRegistryManager_AddRegistry(t *testing.T) {
```
done

byDimasik commented on pkg/helm/oci.go:
```diff
@@ -0,0 +1,111 @@
+func (m *ociRegistryManager) AddRegistry(repoCfg v1beta1.Repository) error {
+	registryURL, err := url.Parse(repoCfg.URL)
+	if err != nil {
+		return fmt.Errorf("can't parse registry URL %s: %w", repoCfg.URL, err)
```
done

byDimasik commented on docs/extensions.md:
```diff
@@ -17,6 +17,9 @@ helm:
     url: https://prometheus-community.github.io/helm-charts
+  - name: oci-registry-with-private-ca
+    url: oci://registry-with-private-ca.com:8080
+    caFile: /path/to/ca.crt
```
please see my comment below - https://github.com/k0sproject/k0s/pull/5901#issuecomment-2941637572

byDimasik commented on pkg/helm/oci.go:
```diff
@@ -0,0 +1,111 @@
+		}),
+	)
+	if err != nil {
+		return fmt.Errorf("can't create registry client for registry %s: %w", repoCfg.URL, err)
```
done

byDimasik commented on pkg/helm/oci.go:
```diff
@@ -0,0 +1,111 @@
+
+// ociRegistryManager is a helper struct for managing OCI registry clients
+type ociRegistryManager struct {
+	knownClients map[string]*registry.Client
```
done

byDimasik commented on inttest/common/bootloosesuite.go:
```diff
@@ -1244,6 +1265,13 @@ func (s *BootlooseSuite) initializeBootlooseClusterInDir(dir string) error {
 	}
 
+	if s.WithRegistry {
+		cfg.Machines = append(cfg.Machines, config.MachineReplicas{
```
fixed - https://github.com/k0sproject/k0s/pull/5901/files#diff-5ce55b33fdea68d04798d1490596b0293a68fc70375cbf254de6f3880dcbf95eR391

byDimasik commented on pkg/helm/oci.go:
```diff
@@ -0,0 +1,111 @@
+		return m.getDefaultRegistryClient()
+	}
+
+	client, exists := m.knownClients[registryURL.Host]
```
please see my comment below - https://github.com/k0sproject/k0s/pull/5901#issuecomment-2941637572

byDimasik commented on pkg/helm/helm.go:
```diff
@@ -220,12 +230,19 @@ func (hc *Commands) InstallChart(ctx context.Context, chartName string, version
+	}
+
 	install := action.NewInstall(cfg)
+	install.SetRegistryClient(registryClient)
```
done

byDimasik commented on pkg/helm/oci.go:
```diff
@@ -0,0 +1,111 @@
+
+// GetRegistryClient returns a registry client for a previously added registry URL or
+// a default registry client if the URL is not known.
+func (m *ociRegistryManager) GetRegistryClient(rawRegistryURL string) (*registry.Client, error) {
```
done

byDimasik commented:
@twz123 I addressed all your comments.

I also removed the cert and key file handling and replaced it with an error that mTLS is not supported for OCI registries. I will proceed with filing a separate enhancement proposal for this.

Now, this change is entirely focused on CA cert

twz123 commented:
I'm fine with not testing or documenting mTLS. However, I'm concerned that we're only matching on the host name and disregarding the path. While this is sufficient for selecting the right CA chain, it could affect future mTLS support. If we leave it as it is now, adding base path matching in the future would be a breaking change.

As a compromise, we could allow repository entries only for OCI registries with the host set and no path elements. This would allow us to opt to support path matching in the future without compromising backwards compatibility.

byDimasik commented:
> I'm fine with not testing or documenting mTLS. However, I'm concerned that we're only matching on the host name and disregarding the path. While this is sufficient for selecting the right CA chain, it could affect future mTLS support. If we leave it as it is now, adding base path matching in the future would be a breaking change.
> 
> As a compromise, we could allow repository entries only for OCI registries with the host set and no path elements. This would allow us to opt to support path matching in the future without compromising backwards compatibility.

@twz123 I've added a path check - https://github.com/k0sproject/k0s/pull/5901/files#diff-bd56e3ef5245c767ecb2d396b681f2d1ea34e530d60151456e1764161984a577R54-R58

Please take another look

twz123 reviewed (dismissed):
Looks good! Kudos for the integration test!

byDimasik commented:
> Looks good! Kudos for the integration test!

@twz123 thank you for the review! 

Now, how do I make sure it's backported to 1.32?

twz123 commented:
Like so. We'll backport it to all actively maintained releases, as long as there's no reason not to do it, starting with 1.33.

byDimasik commented:
> Like so. We'll backport it to all actively maintained releases, as long as there's no reason not to do it, starting with 1.33.

@twz123 
Sounds good! Thank you! Do I get this right that you'll take care of it from here?
I'll keep an eye on it in case you need my help with anything, but as I mentioned earlier, I'll be off next week, and I hope this will be finalized by the release time.

Btw, as promised, here is an issue for potential mTLS support in the future - https://github.com/k0sproject/k0s/issues/5932

twz123 commented:
> Do I get this right that you'll take care of it from here?

Yep! :saluting_face: :upside_down_face:

> Btw, as promised, here is an issue for potential mTLS support in the future - #5932

Nice!

byDimasik commented:
@twz123 Recent changes in the main branch created some merge conflicts that I just resolved, but now the PR needs approval again. Could you take a look?
